text/0002-extracting-necko.md

```diff
@@ -0,0 +1,67 @@
+
+# Analysis: obstacles to embedding
+
+* In its early days, 15+ years ago, Necko was intended to be embeddable; in recent years that has not been an aim of the component leads. Meeting the needs of Gecko and Firefox is more important. (The definition of 'embeddable' here still carries the obligations of Gecko-style initialization.)
+* As noted in the "Embedding Necko" document, Necko depends on Gecko features:
+  - XPCOM and NSPR.
+  - mDNS, `NetUtil.jsm`, and some HTTP functionality requires a JS runtime. This is feasible to eliminate if needed.
+  - Docshell.
+  - If cookies are desired, they must be provided via the nsICookieService interface, which is baked into Necko. One can build Necko without, via `--disable-cookies`, but this is a build-time decision. One could conceivably provide one's own implementation of the cookie service, but the interface requires a stateful synchronous global singleton.
+  - Necko depends on the Mozilla preferences service (libpref) for configuration, which is a privileged singleton that itself depends on the observer service (for startup and profile events) and the directory service (for profile directory and app locations).
+* Because of its dependency on the profile (both for reading state and as a key for the disk cache) and flushing files to disk for prefs, it's not safe to use Necko from two distinct processes simultaneously for the same application.
+* Because of its extensive use of XPCOM singleton patterns, it's not possible to have two independent Neckos in the same process; they need separate processes and separate profile directories. Some independence can be achieved by carefully configuring channels, but it's not isolation.
+* It's not possible to use Necko without a profile service (*e.g.*, to make a network request without touching disk).
+* Parts of Necko are still single-threaded, requiring new channels to be created on the main thread, and dispatching callbacks to the main thread.
+* Necko's init and deinit are managed by observer notifications, which is not ideal for embedding outside of Gecko.
+
+# Proposal
+
+Necko is a very feature-rich C++ networking library that is directly targeted at meeting the needs of single-user, single-main-process, single-profile, interactive applications built on a persistent profile directory in the traditional Mozilla model. It is not well-suited for use outside of the Gecko lifecycle — indeed, almost every aspect is designed for the Gecko world — and would require substantial effort in decoupling to be suited for embedded use. Even if so decoupled, it would still mandate the use of XPCOM and the NSPR.
+
+To extract Necko for reuse outside of Gecko, by non-XPCOM-based applications, would — almost by definition — require coming up with a plan to deal with XPCOM. Given that Necko's data structures, interface definitions, macros, etc. *are* XPCOM's, it seems infeasible to eliminate XPCOM dependency from the module without a total rewrite.
```

Comment:
More input is very welcome!

When I read through the Necko source code, my impression was that practically every method and class somewhere used services, `ns*` classes and data structures, XPCOM errors and results and macros, *et al*. And so my impression is that eliminating what we would call "XPCOM", either narrowly (build and link without use of `/xpcom/`) or broadly (no non-standard C++, including no NSPR and no mozglue and no mfbt), would involve touching almost every line of code.

There are certainly options here — *e.g., * [we could resurrect Oink and use it to rewrite the C++ to something modern](https://brendaneich.com/2006/10/mozilla-2/) — but I still consider that to be a rewrite.

What I consider _not_ to be a total rewrite is being able to grab whole files out of Necko and reuse them with minimal changes: that is, if Necko already carefully encapsulated its use of `nsTArray` and `NS_SUCCEEDED` and `nsCString` and friends, so whole subcomponents were already safe. That doesn't seem to be the case (but I'd be happy to be proved wrong).

Do you think it would be worth splitting this argument into deCOMtamination and "other changes"?